test/ExecuteUpdateTest.cpp

```diff
@@ -0,0 +1,259 @@
+      [&vocab](const VariableToColumnMap& variableColumns,
+               std::vector<SparqlTripleSimpleWithGraph>&& triples,
+               const testing::Matcher<
+                   const std::vector<ExecuteUpdate::TransformedTriple>>&
+                   expectedTransformedTriples,
+               const testing::Matcher<const LocalVocab&>& expectedLocalVocab) {
+        auto [transformedTriples, localVocab] =
+            ExecuteUpdate::transformTriplesTemplate(vocab, variableColumns,
+                                                    std::move(triples));
+        EXPECT_THAT(transformedTriples, expectedTransformedTriples);
+        EXPECT_THAT(localVocab, expectedLocalVocab);
+      };
+  auto expectTransformTriplesTemplateFails =
+      [&vocab](const VariableToColumnMap& variableColumns,
+               std::vector<SparqlTripleSimpleWithGraph>&& triples,
+               const testing::Matcher<const std::string&>& messageMatcher) {
+        AD_EXPECT_THROW_WITH_MESSAGE(
+            ExecuteUpdate::transformTriplesTemplate(vocab, variableColumns,
+                                                    std::move(triples)),
+            messageMatcher);
+      };
+  auto Iri = [](const std::string& iri) {
+    return ad_utility::triple_component::Iri::fromIriref(iri);
+  };
+  auto Literal = [](const std::string& literal) {
+    return ad_utility::triple_component::Literal::fromStringRepresentation(
+        literal);
+  };
+  auto VocabIndex = [](const uint64_t index) {
+    return Id::makeFromVocabIndex(VocabIndex::make(index));
+  };
+  auto LocalVocabIndex = [](const LocalVocabEntry* entry) {
+    return Id::makeFromLocalVocabIndex(entry);
+  };
+  // Transforming an empty vector of template results in no `TransformedTriple`s
+  // and leaves the `LocalVocab` empty.
+  expectTransformTriplesTemplate({}, {}, testing::IsEmpty(),
+                                 testing::IsEmpty());
+  // Resolve a `SparqlTripleSimpleWithGraph` without variables.
+  expectTransformTriplesTemplate(
+      {},
+      {SparqlTripleSimpleWithGraph{Literal("\"foo\""), Iri("<bar>"),
+                                   Literal("\"foo\""),
+                                   SparqlTripleSimpleWithGraph::Graph{}}},
+      testing::ElementsAre(ExecuteUpdate::TransformedTriple{
+          VocabIndex(0), VocabIndex(1), VocabIndex(0), VocabIndex(2)}),
+      testing::IsEmpty());
+  const auto entryNotInIndex = LocalVocabEntry(Iri("<baz>"));
+  // Literals in the template that are not in the index are added to the
+  // `LocalVocab`.
+  expectTransformTriplesTemplate(
+      {},
+      {SparqlTripleSimpleWithGraph{
+          Literal("\"foo\""), Iri("<bar>"), Literal("\"foo\""),
+          SparqlTripleSimpleWithGraph::Graph{::Iri("<baz>")}}},
+      testing::ElementsAre(
+          testing::ElementsAre(VocabIndex(0), VocabIndex(1), VocabIndex(0),
+                               LocalVocabIndex(&entryNotInIndex))),
+      testing::SizeIs(1));
```

Comment:
The correct way to test this is to use `getIndexOrNullopt` on the returned local vocab, and then to *bitwise* compare the IDs via `getBits`, s.t. you actually have the Id stored in the returned local vocab.

Reply:
It is now implemented, but took some effort and time.